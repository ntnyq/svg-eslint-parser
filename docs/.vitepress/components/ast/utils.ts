import { isFunction, isNumber, isObject, isUndefined } from '@ntnyq/utils'
import type { ParentNodeType } from './types'

export function isSVGNode(value: Record<string, any>) {
  return 'type' in value && 'loc' in value && 'range' in value
}

export function getTooltipLabel(
  value: unknown,
  propName?: string,
  parentType?: ParentNodeType,
): string | undefined {
  if (isNumber(value)) {
    switch (parentType) {
      case 'svgNode':
        return propName
    }
  }

  return undefined
}

export function getNodeType(value: unknown): ParentNodeType {
  if (isObject(value) && isSVGNode(value)) {
    return 'svgNode'
  }
  return undefined
}

export function getTypeName(
  value: Record<string, unknown>,
  valueType: ParentNodeType,
): string | undefined {
  switch (valueType) {
    case 'svgNode':
      return String(value.type)

    // extends
  }
  return undefined
}

export function filterProperties(
  key: string,
  value: unknown,
  type: ParentNodeType,
  showTokens?: boolean,
): boolean {
  if (isUndefined(value) || isFunction(value) || key.startsWith('_')) {
    return false
  }

  switch (type) {
    case 'svgNode':
      return key !== 'tokens' || !showTokens
  }

  return true
}
