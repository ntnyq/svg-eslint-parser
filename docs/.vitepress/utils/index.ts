export * from './isIterable'
export * from './scrollIntoView'
