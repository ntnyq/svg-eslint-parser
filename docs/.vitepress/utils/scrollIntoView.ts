/**
 * Scroll given target into view if it's not visible
 * @param target
 */
export function scrollIntoViewIfNeeded(target: Element) {
  const rect = target.getBoundingClientRect()
  const isBelow = rect.top < 0
  const isAbove = rect.bottom > window.innerHeight

  if ((isAbove && isBelow) || rect.height > window.innerHeight) {
    return target.scrollIntoView({
      behavior: 'smooth',
      block: 'start',
      inline: 'start',
    })
  }

  // outside from the view
  if (isAbove || isBelow) {
    return target.scrollIntoView({
      behavior: 'smooth',
      block: 'center',
      inline: 'center',
    })
  }
}
