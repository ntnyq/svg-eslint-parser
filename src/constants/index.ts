export * from './char'
export * from './parse'
export * from './nodeTypes'
export * from './tokenTypes'
export * from './svgElements'
export * from './tokenizerContextTypes'
export * from './constructTreeContextTypes'
