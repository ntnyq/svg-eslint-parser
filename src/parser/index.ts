export * from './error'
export * from './parse'
export * from './traverse'
export * from './parseForESLint'
