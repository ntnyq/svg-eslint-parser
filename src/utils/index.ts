export * from './firstLast'
export * from './cloneRange'
export * from './initIfNone'
export * from './getLineInfo'
export * from './clearParent'
export * from './isWhitespace'
export * from './cloneLocation'
export * from './updateNodeEnd'
export * from './createNodeFrom'
export * from './parseOpenTagName'
export * from './parseCloseTagName'
export * from './calculateTokenLocation'
export * from './calculateTokenPosition'
