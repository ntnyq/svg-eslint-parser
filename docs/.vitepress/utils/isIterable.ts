import { isFunction, isObject } from '@ntnyq/utils'

export function isIterable(value: unknown) {
  return isObject(value) && Symbol.iterator in value && isFunction(value[Symbol.iterator])
}
