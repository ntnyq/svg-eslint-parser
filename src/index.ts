import { meta } from './meta'
import { ParseError, parseForESLint } from './parser'
import { visitorKeys } from './visitorKeys'
import type { Options } from './types'
import type * as AST from './types'

export function parseSVG(code: string, options: Options = {}) {
  return parseForESLint(code, options).ast
}

export const name = meta.name
export const VisitorKeys = visitorKeys

export const parse = parseSVG

export { meta }
export { ParseError }
export { parseForESLint }

export type { AST }

export * from './types'
export * from './constants'
