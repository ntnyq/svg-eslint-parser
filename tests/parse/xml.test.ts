import { unindent as $ } from '@ntnyq/utils'
import { expect, it } from 'vitest'
import { parseForESLint } from '../../src'

const SVG_SOURCE = $`
  <?xml version="1.0" encoding="UTF-8" standalone="no"?>
`

it('should parse', () => {
  const ast = parseForESLint(SVG_SOURCE).ast

  expect(ast).toMatchSnapshot()
  expect(ast.body).toMatchInlineSnapshot(`
    [
      {
        "children": [
          {
            "loc": {
              "end": {
                "column": 54,
                "line": 1,
              },
              "start": {
                "column": 0,
                "line": 1,
              },
            },
            "range": [
              0,
              54,
            ],
            "type": "Text",
            "value": "<?xml version="1.0" encoding="UTF-8" standalone="no"?>",
          },
        ],
        "loc": {
          "end": {
            "column": 54,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "range": [
          0,
          54,
        ],
        "type": "Document",
      },
    ]
  `)
})
