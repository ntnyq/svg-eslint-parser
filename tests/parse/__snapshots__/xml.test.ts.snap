// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should parse 1`] = `
{
  "body": [
    {
      "children": [
        {
          "loc": {
            "end": {
              "column": 54,
              "line": 1,
            },
            "start": {
              "column": 0,
              "line": 1,
            },
          },
          "range": [
            0,
            54,
          ],
          "type": "Text",
          "value": "<?xml version="1.0" encoding="UTF-8" standalone="no"?>",
        },
      ],
      "loc": {
        "end": {
          "column": 54,
          "line": 1,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      "range": [
        0,
        54,
      ],
      "type": "Document",
    },
  ],
  "comments": [],
  "loc": {
    "end": {
      "column": 54,
      "line": 1,
    },
    "start": {
      "column": 0,
      "line": 1,
    },
  },
  "range": [
    0,
    54,
  ],
  "tokens": [
    {
      "loc": {
        "end": {
          "column": 54,
          "line": 1,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      "range": [
        0,
        54,
      ],
      "type": "Text",
      "value": "<?xml version="1.0" encoding="UTF-8" standalone="no"?>",
    },
  ],
  "type": "Program",
}
`;
